pageserver/src/tenant/timeline/layer_manager.rs

```diff
@@ -337,8 +337,23 @@ impl OpenLayerManager {
         compact_to: &[ResidentLayer],
         metrics: &TimelineMetrics,
     ) {
-        // We can simply reuse compact l0 logic. Use a different function name to indicate a different type of layer map modification.
-        self.finish_compact_l0(compact_from, compact_to, metrics)
+        // gc-compaction could contain layer rewrites. We need to delete the old layers and insert the new ones.
```

Comment:
Can you model this function as `OpenLayerManager::rewrite_layers` + addition of any new layers which aren't re-writes. You also need to carry over the visibility hint, which the mentioned function does and this function doesn't.

Reply:
resolved in https://github.com/neondatabase/neon/pull/10430/commits/ff373ddab28e8cd1b6e1fb107bfdf54f973b6f4a

Reply:
I would update `rewrite_layers` to contain the extra functionality you need of adding new layers in the same batch of updates (or make it return the batch so that you can append to it).

Normally, I don't mind duplication too much, but this this code feels a bit brittle. If we find a but in it, it's likely we'll forge to update both "call sites".

Reply:
consolidated them at https://github.com/neondatabase/neon/pull/10430/commits/1bb54ef9b7f7b32f17be66ac12e576b52554a911 :)